### ISSUE TITLE
Implement (finally!) Podani's extension to Gower's coefficient for ordinal data

### ISSUE THREAD

**Issue by gavinsimpson** [gavinsimpson/analogue#15]
I started this a while back and never got round to finishing it off. Basically need to handle ordinal values as per Podani's 1999 extension of Gower's coefficient to ordinal data. Currently `distance()` handles ordinal values as nominal values.

Needs two versions
1. non-metric: this is Podani's equation 2b, which requires tabulations of ranks (number of observations of same rank for each rank for each variable).
2. metric: this is Podani's equation 3, which requires only the absolute difference in ranks of samples _i_ and _j_, standardised by the difference between the largest and smallest rank values for each variable _k_.

Option 2. is easy to implement as we can just convert those ordinal variables to ranks and then treat them like quantitative variables.

Option 1. requires a little more thought as to how to handle passing in of the rank meta information (the _Ti_ values and the _Ti,max_ information). Don't want to overly complicate the C code when perfectly good R functions can do this for us.

**Comment by jarioksa** [comment]
Emmanuel Laliberté has implemented this in `gowdis()` function in his **FD** package. There is **C** code with nested  `switch`es.

*Taxon* asked me to send 20 bugs to see Podani's paper. I didn't. So I can't check his equations to think how to implement this nicely. However, the rank transformation (with average ranks for ties)  and range adjustment can be done in **R** before calling **C**. The information like the counts per rank and ranges of ranks can be collected within the **C** code. The only new argument in the **C** call is the type of the variable because numbers coming from different kinds of variables are treated differently within the code.

**Comment by gavinsimpson** [comment]
One issue I see with the **FD** implementation is that for all combinations of *j* and *k* sample pairings it is computing the *Ti*. The C code loops over the observations for each combination of *j* and *k* and each variable *i*. It seems wasteful to me to keep doing this individually for all pairs rather than once for all ranks.

But: the tabulated counts per rank will be a ragged array any time there are different numbers of ranks per variable. And as we can't easily have a ragged array (or any special R types - like lists) via the `.C` interface, this presents a problem.

At the risk of chewing up more memory, one option would be compute the ranks and then do an `ave()` like operation such that we have a matrix of dimension *N* by *I* (*N* number of samples, *I* number of variables) where the entries are the number of observations with the same rank as the *n* = {1,...,N} th observation for the *i*th variable.

For example:

```r
set.seed(1)
df <- data.frame(x = ordered(sample(1:3, 10, replace = TRUE)),
                 y =  ordered(sample(LETTERS[1:4], 10, replace = TRUE),
                              levels = LETTERS[1:4]))
T <- sapply(df, function(x) tabulate(x)[x])
```

which gives

```r
> T
      x y
 [1,] 3 2
 [2,] 4 2
 [3,] 4 2
 [4,] 3 3
 [5,] 3 3
 [6,] 3 3
 [7,] 3 2
 [8,] 4 3
 [9,] 4 3
[10,] 3 3
```

That leaves the *Ti,max* and *Ti,min* bit. This can be done ahead of time in R too and even precompute the entire `(Ti,max - 1)/2 - (Ti,min - 1)/2` bit to pass in as we do now for `R` (the ranges of each variable). To be efficient, we should do this from the `tabulate()` output used to create `T` so rather than loop twice do

```r
nc <- NCOL(df)
nr <- NROW(df)
T <- matrix(ncol = nc, nrow = nr)
Trange <- numeric(length = nc)
for (i in seq_len(NCOL(df))) {
  tab <- tabulate(df[, i])
  T[, i] <- tab[df[, i]]
  tminmax <- (tab[c(1, length(tab))] - 1) / 2
  Trange[i] <- tminmax[2] - tminmax[1]
}
```

`T` is unchanged and `Trange` is now

```r
> Trange
[1] 0.0 0.5
```

Now I can use the (in C) row indices `i` and `j` to index into `T` just as I would `x` (the matrix of observations cast as `double(x)`), and index into `Trange` just as I do with current argument `R`.

If you have any thoughts @jarioksa, on this or another way to do it, they would be gladly received.

**Comment by jarioksa** [comment]
Gav, I worked out an idea of doing this in 792f2e6462ae98b9cc9e452a129a92232107a49d . This provides a **C** implementation of the actual loop leaving most of the responsibility of fixing data to the **R** side. The basic idea is that Podani's ordinal method can be rewritten as 

(`Manhattan` - `Tiefix`) / `scale`

Where `scale`  is the denominator in Podani 1999 eq 2b. This `scale` is found from the range of the variable and number of items tied to minimum and maximum. It is the constant for each variable so that it can be supplied as the input. His metric ordinal case is just `Manhattan` / `scale`  where `scale` is the range of the variable (that again can be calculated in the calling programme and supplied as input).  The eq 2b can be rearranged as

`Manhattan` / `scale` - `Tiefix` / `scale`

Here `Tiefix` is also a dissimilarity-like object and can be calculated in the same loop if we add columns giving information on the number of ties to the data matrix or call the function another time for ties matrix.

A naïve way of getting the ties would be something like this:
```R
library(vegan)
data(dune.env)
rowSums(with(dune.env, outer(Moisture, Moisture, "==")))
 # [1] 7 7 4 4 7 7 7 7 2 4 7 2 7 7 7 7 4 7 7 7
```

The comments are extensive in 792f2e6462ae98b9cc9e452a129a92232107a49d and it may be best to inspect my ideas there. It is now in separate file to minimize disturbance. 

Looking from **C** side, most variable types do not make much difference. It is only about how they are flagged, how they are tranformed, and how their scale was defined somewhere in the outer world.

**Comment by jarioksa** [comment]
I added R code to cast input data into form that can be handled in my suggested **C** code (d01af00e8fbb964dda2d739cc7ebbe6fd86f04fd). All data wrangling is done in **R**, and the **C** interface would be just as simple as in ordinary distance functions. 

There are two disjunct pieces: **R** code to make the data and supplementary vectors, and **C** code to analyse these data. What is missing is the piece that combines these and actually calculates the distances. These are proof-of-the-concept cases. Use them if you think they are usable.

**Comment by jarioksa** [comment]
I added **C** driver and  `gowerdistance` function now returns a vector of dissimilarities. They have no class nor any other attributes and are not recognized as distances. Adding class and attributes would solve that, but I think that is not needed for testing: element by element the vector is equal to a similar distance object.

This still is a proof-of-the-concept. In particular the **R** interface is fragile, and for instance, will not cope with `NA`. The **C** is intended as readier for use (except the driver which is _ad hoc_).

**Comment by jarioksa** [comment]
You asked about `vtype`  `T` in `daisy()`. This is from their **Fortran** file:
```
c       vtype(j) is the type of variable j:
c              = 1 (A) for an Asymmetric binary variable
c              = 2 (S) for a  Symmetric  binary variable
c              = 3 (N) for a  Nominal  variable
c              = 4 (O) for an Ordinal  variable
c              = 5 (I) for an Interval variable (additive)
c              = 6 (T) for a  raTio    variable (log transformed)
```
Gower (1971) did not have but three variable types: A, N & I. Others are extensions to his work, probably by Kaufman & Rousseeuw. `daisy` uses internal codes `1..nlevels(A)` for ordered factors (vtype O) instead of ranks, which may be Podani's innovation. Automatic log transformation for ratio-scale variables (vtype T) probably is a `daisy` innovation.

**Comment by gavinsimpson** [comment]
Thanks @jarioksa. My C code is broken and not returning the right distances. I haven't figured out where I made the error yet. I'll take a look at you branch later today.

**Comment by jarioksa** [comment]
Actually my simple idea seems not to work... I did not observe that the tiefix should only be applied if  `fabs(x[i1] - x[i2]) > 0`. I got to think how to do this now... This means that the Podani non-metrix now fails, but all other cases seem to work OK. The fix is doable, but needs a bit more complicated interface.

**Comment by jarioksa** [comment]
After 6d095d06dfe97a6d0c my fork gives similar results as `FD::gowdis()` also for Podani ties. At least in first tests. This is a quick hack: I try to find a neater solution.

**Comment by jarioksa** [comment]
I ran some more tests, and my function seems to give numerically identical results to `FD::gowdis()` with all options for ordered variables. I suggest you take that as the base tool. The **R** interface also works, but needs a lot of concern for `NA` etc.

Basically I have four different cases of handling contribution to the dissimilarity in the **C** code:

1. Matching coefficient that is used for nominal variables.
2. Manhattan with Podani's tie fix that is optionally used for ordered factors.
3. Manhattan distance which is used for continuous variables and optionally for ordered factors. The **C** code is ignorant to coding of ordered factors which happens in **R** code, but whatever the coding is, the result is handled like continuous numeric variables.
4. "Jaccard-like" or "Gower distance" which is Manhattan ignoring double zeros that is optionally used for non-negative continuous variables and binary (logical) variables.

So this is what happens in **C**. Another issue is how different kind of input data are mapped into these cases, and how ordered factors etc are coded. That will happen in **R**. The **C** code itself is quite ignorant of original data types: it only works like requested.

**Comment by jarioksa** [comment]
Your code had two problems that I have fixed in [my branch](https://github.com/jarioksa/analogue/tree/Trange-fix):

1. `Trange` was wrongly calculated in the **R** code (debb362b51)
2. There was a stray `break;` in the **C** code. This causes skipping the rest of the step, including update of `wsum` (b4a86a2698f)

After these fixes, your issue-#15 branch gives identical results to `FD::gowdis` or `gowerdistance` in my gower-podani branch.

**Comment by jarioksa** [comment]
The following fails because `mite.env` in **vegan** has only one ordered factor:
```R
> distance(mite.env, method="mixed")
Error in T[, ordinal][, i] <- tab[orderedx[, i]] : 
  incorrect number of subscripts on matrix
```
Fix is in 5cc4390b (against issue-#15 branch). 

My version of the same (in my gower-podani branch) uses completely different approach.

The following failed because `BCI.env`  in **vegan** has some constant variables:
```R
distance(BCI.env, method="mixed")
```
Fix is in 9dea984285ccda233b5990072a30978b67e739f2 (aisnt issue-#15 branch). 

Function divides with `R`ange and constant variables give contribution `0/0` or `NaN`. Solution was to replace zero-range with a positive number so that `0/R` gives `0`. Keeping variables (instead of removing) will influence sum of weights and so distances. `FD::gowdis()` and my branch fail with the same data for the same reason.
